tests/Feature/CityControllerTest.php

```diff
@@ -0,0 +1,170 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tests\Feature;
+
+use App\Models\City;
+use App\Models\Country;
+use Generator;
+use Inertia\Testing\AssertableInertia as Assert;
+use Tests\TestCase;
+
+class CityControllerTest extends TestCase
+{
+    public function testCitiesViewCanBeRendered(): void
+    {
+        City::factory()->count(3)->create();
+
+        $this->get("/cities")->assertInertia(
+            fn(Assert $page) => $page
+                ->component(value: "Cities")
+                ->has("cities.data", 3),
+        );
+    }
+
+    public function testCityCanBeCreated(): void
+    {
+        $country = Country::factory()->create();
+
+        $city = [
+            "name" => "Legnica",
+            "latitude" => 44.543,
+            "longitude" => -43.122,
+            "country_id" => $country->id,
+        ];
+
+        $this->post(route(name:"cities.store"), data: $city);
+
+        $this->assertDatabaseHas(table:"cities", data: $city);
+    }
+
+    public function testCityCannotBeCreatedBecauseFieldsAlreadyExist(): void
+    {
+        $country = Country::factory()->create();
+
+        $city = [
+            "name" => "Legnica",
+            "latitude" => 44.543,
+            "longitude" => -43.122,
+            "country_id" => $country->id,
+        ];
+
+        City::query()->create([
+            "name" => $city["name"],
+            "latitude" => -55.54323,
+            "longitude" => 42.3721,
+            "country_id" => $country->id,
+        ])->toArray();
+
+        $this->post(route(name:"cities.store"), data: $city);
+
+        $this->assertDatabaseMissing(table: "cities", data: $city);
+    }
+
+    /**
+     * @dataProvider invalidCityDataProvider
+     */
+    public function testCityCannotBeCreatedWithInvalidData(array $data, array $expectedErrors): void
+    {
+        $response = $this->post(route("cities.store"), $data);
+
+        $response->assertSessionHasErrors($expectedErrors);
+    }
```

Comment:
Its more readable test for me. What do u think @EwelinaLasowy?
```suggestion
    public function testCityCannotBeCreatedWithInvalidData(): void
    {
        $data = [
            "data" => [
                    "name" => null,
                    "longitude" => null,
                    "latitude" => null,
                ]
            ];
            
        $response = $this->post(route("cities.store"), $data);

        $response->assertSessionHasErrors(["name", "latitude", "longitude"]);
    }
```

Reply:
It uses DataProvider to creating and updating.

Reply:
I know but that test for you is readable?

Reply:
It's recommended by @mtracz.